sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeCreator.scala

```diff
@@ -132,50 +153,80 @@ case class CreateMap(children: Seq[Expression]) extends Expression {
     new ArrayBasedMapData(new GenericArrayData(keyArray), new GenericArrayData(valueArray))
   }
 
+  private def getAccessors(ctx: CodegenContext, dt: DataType, array: String,
+      isPrimitive : Boolean, size: Int): (String, String, String) = {
+    if (!isPrimitive) {
+      val arrayClass = classOf[GenericArrayData].getName
+      ctx.addMutableState("Object[]", array, s"this.$array = null;")
+      (s"new $arrayClass($array)",
+        s"$array = new Object[${size}];", s"this.$array = null;")
+    } else {
+      val unsafeArrayClass = classOf[UnsafeArrayData].getName
+      val javaDataType = ctx.javaType(dt)
+      ctx.addMutableState(s"${javaDataType}[]", array,
+        s"this.$array = new ${javaDataType}[${size}];")
+      (s"$unsafeArrayClass.fromPrimitiveArray($array)", "", "")
```

Comment:
I left the conventional approach for non-primitive cases. Now, I tried to allocate an object only once for non-primitive cases, too.